### PR TITLE
feat: Introduce Google Cloud Storage Implementation

### PR DESCRIPTION
**Pull request by Xuanwo**
### Description

Part of https://github.com/quickwit-oss/quickwit/issues/4236

This PR implements:

- a basic opendal storage implementation (not exposed)
- Google Cloud Storage Implementation (exposed via config)

There are many TODOs to finish, but I feel like it's better to make it working first.

### How was this PR tested?

Added `google_cloud_storage_test_suite`

### CODE REVIEW

File: quickwit/quickwit-storage/src/opendal_storage.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
+// Copyright (C) 2023 Quickwit, Inc.
```
Comment (etolbakov) [inline quickwit/quickwit-storage/src/opendal_storage.rs:1]: nit: but I guess this needs an `inc` 😅

File: quickwit/Cargo.toml [hunk 2]
```diff
@@ -118,6 +118,7 @@ num_cpus = "1"
 numfmt = "1.1.1"
 once_cell = "1"
 oneshot = "0.1.5"
+opendal = { version ="0.44", default-features = false }
```
Comment (etolbakov) [inline quickwit/Cargo.toml:121]: not sure I understood why `opendal` is declared here. 
Will it be enough to define it only in the `storage` module?

Comment (Xuanwo) [reply quickwit/Cargo.toml:121]: > not sure I understood why `opendal` is declared here.

I found most deps listed in `quickwit` are using `workspace = true` like:

https://github.com/quickwit-oss/quickwit/blob/c4e449ff941c6263386a00cf94c347e853e3f030/quickwit/quickwit-storage/Cargo.toml#L38-L46

I have no preference. I'm happy to change it if you'd like.

File: quickwit/quickwit-storage/src/opendal_storage.rs (hunk 1 above)
Comment (Xuanwo) [reply quickwit/quickwit-storage/src/opendal_storage.rs:1]: Addressed.

File: quickwit/Cargo.toml (hunk 2 above)
Comment (etolbakov) [reply quickwit/Cargo.toml:121]: fair point, let's leave it then 👍

**Comment by Xuanwo** [comment]
Hi @etolbakov, what are your thoughts on the current plan? Should we first merge the OpenDAL storage and then add GCS storage? For GCS, integration with configurations and settings is necessary. Once GCS is operational, we can incorporate metrics and other utilities similar to those provided by S3 or FS.

**Comment by etolbakov** [comment]
Hey @Xuanwo 
Actually I was doing some work on this ticket in the background, to my shame I didn't assign/mention that anyhow.
My implementation strategy was the opposite: using `opendal` exactly for gcs integration, make sure it fits nicely in the project, then make it generic and potentially migrate other storage layers. 
But I totally see where you are coming from. So we may join our efforts.
> but I feel like it's better to make it working first.
Could you please elaborate a bit about what do you mean by "working first".  Is it a simple integration test that we can draft?

Feel free to reach out in the discord if that's fine with you

**Comment by Xuanwo** [comment]
> Feel free to reach out in the discord if that's fine with you

Great, moved to https://discord.com/channels/908281611840282624/908281611840282627/1192630218482012180

**Comment by Xuanwo** [comment]
Hi @etolbakov, could you share your Discord ID? I'd like to invite you to our discussion thread: https://discord.com/channels/908281611840282624/1192630218482012180

**Comment by fulmicoton** [comment]
@Xuanwo I like the approach: using a generic opendal storage, and adding a factory for googlefs.
I should have time tomorrow to properly review.

File: quickwit/quickwit-storage/src/opendal_storage/base.rs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+    /// # TODO
+    ///
+    /// We can implement something like `multipart_policy` determine whether to use copy.
+    /// If the payload is small enough, we can call `op.write()` at once.
+    async fn put(&self, path: &Path, payload: Box<dyn PutPayload>) -> StorageResult<()> {
+        let path = path.as_os_str().to_string_lossy();
+        let mut r = payload.byte_stream().await?.into_async_read();
+
+        let mut w = self.op.writer_with(&path).buffer(8 * 1024 * 1024).await?;
```
Comment (etolbakov) [inline quickwit/quickwit-storage/src/opendal_storage/base.rs:84]: feel free to use `ByteSize::mb(8)` here

File: quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs [hunk 4]
```diff
@@ -0,0 +1,124 @@
+// Copyright (C) 2023 Quickwit, Inc.
```
Comment (etolbakov) [inline quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs:1]:
```suggestion
// Copyright (C) 2024 Quickwit, Inc.
```
nit: same `inc` 🤣

File: quickwit/quickwit-storage/src/opendal_storage/base.rs (hunk 3 above)
Comment (Xuanwo) [reply quickwit/quickwit-storage/src/opendal_storage/base.rs]: Addressed.

File: quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs (hunk 4 above)
Comment (Xuanwo) [reply quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs]: Addressed.

File: quickwit/Cargo.toml (hunk 2 above)
Comment (fulmicoton) [inline quickwit/Cargo.toml:121]: can we feature gate this to make the development a tiny bit lighter?
(You can mimick what is done for azure.)

File: quickwit/quickwit-config/src/storage_config.rs [hunk 5]
```diff
@@ -374,6 +391,26 @@ pub struct FileStorageConfig;
...
+    pub credential_path: Option<String>,
+}
+
+impl GoogleCloudStorageConfig {
+    pub const GOOGLE_CLOUD_STORAGE_CREDENTIAL_PATH_ENV_VAR: &'static str = "QW_GOOGLE_CLOUD_STORAGE_CREDENTIAL_PATH";
+
+    /// Attempts to find the credential path in the environment variable
+    /// `QW_GOOGLE_CLOUD_STORAGE_CREDENTIAL_PATH` or the config.
+    pub fn resolve_credential_path(&self) -> Option<String> {
+        env::var(Self::GOOGLE_CLOUD_STORAGE_CREDENTIAL_PATH_ENV_VAR)
```
Comment (fulmicoton) [inline quickwit/quickwit-config/src/storage_config.rs:408]: I think we tend to give priority to configuration, and when absent, inspect environment variables.

Comment (Xuanwo) [reply quickwit/quickwit-config/src/storage_config.rs:408]: Oops, I copied those logic from azure storage.

File: quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+/// TODO: maybe we can also support `gs://bucket`?
+fn parse_google_uri(uri: &Uri) -> Option<(String, PathBuf)> {
+    // Ex: google://bucket/prefix.
+    static URI_PTN: OnceCell<Regex> = OnceCell::new();
+
+    let captures = URI_PTN
+        .get_or_init(|| {
+            Regex::new(r"google(\+[^:]+)?://(?P<bucket>[^/]+)(/(?P<prefix>.+))?")
```
Comment (fulmicoton) [inline quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs:85]: I think `gs` is the usual protocol for google cloud storage.
Have you seen google:// somewhere?

Comment (fulmicoton) [reply quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs:85]: I think your regex does not allow `google://bucket/`.
Shouldn't we allow it?

File: quickwit/quickwit-storage/src/opendal_storage/base.rs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+        Ok(Self { uri, op })
+    }
+}
+
+#[async_trait]
+impl Storage for OpendalStorage {
+    async fn check_connectivity(&self) -> anyhow::Result<()> {
+        self.op.check().await?;
+
```
Comment (fulmicoton) [inline quickwit/quickwit-storage/src/opendal_storage/base.rs:73]:
```suggestion
```

File: quickwit/quickwit-storage/src/opendal_storage/base.rs [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+impl OpendalStorage {
+    /// Create a new google cloud storage.
+    pub fn new_google_cloud_storage(
+        uri: Uri,
+        cfg: opendal::services::Gcs,
+    ) -> Result<Self, StorageResolverError> {
+        let op = Operator::new(cfg)?.finish();
+
```
Comment (fulmicoton) [inline quickwit/quickwit-storage/src/opendal_storage/base.rs:64]:
```suggestion
```

File: quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs (hunk 6 above)
Comment (Xuanwo) [reply quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs:85]: Those logic copied from azure and check by `test_parse_google_uri`. `google://bucket/` should work.

Comment (Xuanwo) [reply quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs]: > I think `gs` is the usual protocol for google cloud storage.

Addressed.

File: quickwit/quickwit-config/src/storage_config.rs (hunk 5 above)
Comment (Xuanwo) [reply quickwit/quickwit-config/src/storage_config.rs:408]: The current logic aligns with S3 and Azure. Should I open a new issue to investigate this behavior?

File: quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+#[async_trait]
+impl StorageFactory for GoogleCloudStorageFactory {
+    fn backend(&self) -> StorageBackend {
+        StorageBackend::Google
+    }
+
+    async fn resolve(&self, uri: &Uri) -> Result<Arc<dyn Storage>, StorageResolverError> {
+        let storage = from_uri(&self.storage_config, uri)?;
```
Comment (fulmicoton) [inline quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs:45]: I think to get it to work you also need to add a google protocol in `uri.rs`, and add it in `storage_resolver.rs`

File: quickwit/quickwit-storage/src/opendal_storage/base.rs [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+        Ok(())
+    }
+
+    /// # TODO
+    ///
+    /// We can implement something like `multipart_policy` determine whether to use copy.
+    /// If the payload is small enough, we can call `op.write()` at once.
+    async fn put(&self, path: &Path, payload: Box<dyn PutPayload>) -> StorageResult<()> {
+        let path = path.as_os_str().to_string_lossy();
+        let mut r = payload.byte_stream().await?.into_async_read();
```
Comment (fulmicoton) [inline quickwit/quickwit-storage/src/opendal_storage/base.rs:81]: no one letter variable please.

File: quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs (hunk 9 above)
Comment (Xuanwo) [reply quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs:45]: Working on it.

File: quickwit/quickwit-config/src/storage_config.rs (hunk 5 above)
Comment (fulmicoton) [reply quickwit/quickwit-config/src/storage_config.rs:408]: Oh ok! Let it that way. I'll open the ticket. Thanks.

File: quickwit/quickwit-storage/src/opendal_storage/base.rs [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+#[async_trait]
+impl Storage for OpendalStorage {
+    async fn check_connectivity(&self) -> anyhow::Result<()> {
+        self.op.check().await?;
+        Ok(())
+    }
+
+    /// # TODO
+    ///
+    /// We can implement something like `multipart_policy` determine whether to use copy.
```
Comment (fulmicoton) [inline quickwit/quickwit-storage/src/opendal_storage/base.rs:77]: So opendal supports multipart upload? What does the API look like?

File: quickwit/quickwit-storage/src/opendal_storage/base.rs [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    async fn get_slice_stream(
+        &self,
+        path: &Path,
+        range: Range<usize>,
+    ) -> StorageResult<Box<dyn AsyncRead + Send + Unpin>> {
+        let path = path.as_os_str().to_string_lossy();
+        let range = range.start as u64..range.end as u64;
+        let r = self.op.reader_with(&path).range(range).await?;
```
Comment (fulmicoton) [inline quickwit/quickwit-storage/src/opendal_storage/base.rs:117]: no one letter variable

**Review by fulmicoton — CHANGES_REQUESTED** [review]
Do you know a way to mock gfs for our CI? Also have you seen the storage testsuite?

File: quickwit/quickwit-storage/src/opendal_storage/base.rs [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+        Ok(Box::new(r))
+    }
+
+    async fn get_all(&self, path: &Path) -> StorageResult<OwnedBytes> {
+        let path = path.as_os_str().to_string_lossy();
+        let bs = self.op.read(&path).await?;
```
[the remark this answers is not in the document]
Comment (Xuanwo) [reply quickwit/quickwit-storage/src/opendal_storage/base.rs:124]: > what does bs stands for?

I will replace 'bs' with more descriptive terms and update previous comments for clarity.

File: quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs (hunk 9 above)
Comment (Xuanwo) [reply quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs:52]: Addressed.

File: quickwit/quickwit-storage/src/opendal_storage/base.rs (hunk 11 above)
Comment (Xuanwo) [reply quickwit/quickwit-storage/src/opendal_storage/base.rs:77]: > So opendal supports multipart upload? What does the API look like?

OpenDAL provides multipart support through the `Writer` interface. Each invocation of `writer.write()` triggers a multipart upload. Users can adjust internal behavior as follows:

- `op.writer_with(path).buffer(buffer_size)` set the buffer size of each part.
- `op.writer_with(path).concurrent(concurrent)` set the concurrent number to part uploading.

**Comment by Xuanwo** [comment]
> Do you know a way to mock gfs for our CI? Also have you seen the storage testsuite?

I have added [fake-gcs-server](https://github.com/fsouza/fake-gcs-server) in CI and docker compose, but I don't know how get it started.

**Comment by Xuanwo** [comment]
> I have added [fake-gcs-server](https://github.com/fsouza/fake-gcs-server) in CI and docker compose, but I don't know how get it started.

Ok, I got the right place. Let me add them.

File: .github/workflows/coverage.yml [hunk 14]
```diff
@@ -115,6 +115,14 @@ jobs:
           username: ${{ secrets.DOCKERHUB_USERNAME }}
           password: ${{ secrets.DOCKERHUB_ACCESS_TOKEN }}
 
+      fake-gcs-server:
+        image: fsouza/fake-gcs-server:1.47.7
+        ports:
+          - "4443:4443"
+        credentials:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_ACCESS_TOKEN }}
```
Comment (fulmicoton) [inline .github/workflows/coverage.yml:124]: Do we need the credentials here? 
I see you just copied the other stuff but I think this is an error on our side too.

Comment (Xuanwo) [reply .github/workflows/coverage.yml:124]: > Do we need the credentials here?

I do think we don't need credentials to pull service containers. Let me remove them.

Comment (Xuanwo) [reply .github/workflows/coverage.yml]: Removed.

**Comment by Xuanwo** [comment]
I have added integration tests for gcs:

```rust
test google_cloud_storage_test_suite ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.19s
```

The mock gcs server doesn't properly support bulk deletion, so I've disabled that feature. However, I'm confident it functions correctly on the actual gcs.

File: quickwit/Cargo.toml [hunk 15]
```diff
@@ -247,6 +247,9 @@ azure_storage_blobs = { version = "0.13.0", default-features = false, features =
   "enable_reqwest_rustls",
 ] }
 
+opendal = { version ="0.44", default-features = false }
+reqsign = { version ="0.14", default-features = false }
```
Comment (fulmicoton) [inline quickwit/Cargo.toml:251]:
```suggestion
opendal = { version = "0.44", default-features = false }
reqsign = { version = "0.14", default-features = false }
```

File: quickwit/quickwit-storage/src/lib.rs [hunk 16]
```diff
@@ -320,9 +327,14 @@ pub(crate) mod test_suite {
         test_write_and_delete(storage)
             .await
             .context("write_and_delete")?;
-        test_write_and_bulk_delete(storage)
-            .await
-            .context("write_and_bulk_delete")?;
+        // Fake GCS Server doesn't support bulk delete correctly.
+        // ref: <https://github.com/fsouza/fake-gcs-server/issues/1443>
+        #[cfg(not(feature = "google"))]
```
Comment (fulmicoton) [inline quickwit/quickwit-storage/src/lib.rs:332]: Let's not do the check using the compiler flag, but instead dynamically inspect the storage.
As is, your code is removing this test for azure too in our CI.

(these test are running in the confusingly named "coverage CI" that is triggered after each merge to main. That build takes a bit more time than the regular PR CI and includes azure etc.)

File: quickwit/quickwit-storage/src/opendal_storage/base.rs (hunk 11 above)
Comment (fulmicoton) [reply quickwit/quickwit-storage/src/opendal_storage/base.rs:77]: awesome thank you!

File: quickwit/quickwit-storage/src/opendal_storage/base.rs [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+        let path = path.as_os_str().to_string_lossy();
+        self.op.delete(&path).await?;
+
+        Ok(())
+    }
+
+    async fn bulk_delete<'a>(&self, paths: &[&'a Path]) -> Result<(), BulkDeleteError> {
+        let paths = paths
+            .iter()
+            .map(|v| v.as_os_str().to_string_lossy().to_string())
```
Comment (fulmicoton) [inline quickwit/quickwit-storage/src/opendal_storage/base.rs:139]:
```suggestion
            .map(|path| path.as_os_str().to_string_lossy().to_string())
```

File: quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+    Ok(store)
+}
+
+fn parse_google_uri(uri: &Uri) -> Option<(String, PathBuf)> {
+    // Ex: gs://bucket/prefix.
+    static URI_PTN: OnceCell<Regex> = OnceCell::new();
+
+    let captures = URI_PTN
+        .get_or_init(|| {
+            Regex::new(r"gs(\+[^:]+)?://(?P<bucket>[^/]+)(/(?P<prefix>.+))?")
```
Comment (fulmicoton) [inline quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs:122]: nitpick. 

`gs://bucket/` works thanks to partial matching.

Maybe we could tweak it into:
```suggestion
            Regex::new(r"gs(\+[^:]+)?://(?P<bucket>[^/]+)(/(?P<prefix>.+)?)?$")
```
or
```suggestion
            Regex::new(r"gs(\+[^:]+)?://(?P<bucket>[^/]+)(/(?P<prefix>.*))?$")
```
(note the $ at the end)

File: quickwit/quickwit-storage/src/lib.rs (hunk 16 above)
Comment (Xuanwo) [reply quickwit/quickwit-storage/src/lib.rs]: Addresed, please review again

File: quickwit/quickwit-config/src/storage_config.rs [hunk 19]
```diff
@@ -157,13 +168,14 @@ pub enum StorageConfig {
...
     S3(S3StorageConfig),
+    Google(GoogleCloudStorageConfig),
 }
 
 impl StorageConfig {
     pub fn redact(&mut self) {
         match self {
             Self::Azure(azure_storage_config) => azure_storage_config.redact(),
-            Self::File(_) | Self::Ram(_) => {}
+            Self::File(_) | Self::Ram(_) | Self::Google(_) => {}
```
Comment (etolbakov) [inline quickwit/quickwit-config/src/storage_config.rs:178]: I curious why `Self::Google` does not follow  `Self::S3` (and `Self::Azure`) logic with redact?

**Review by etolbakov — APPROVED** [review]
@Xuanwo looks wonderful! I don't have anything to suggest!

File: quickwit/quickwit-storage/Cargo.toml [hunk 20]
```diff
@@ -70,11 +74,19 @@ azure = [
   "azure_storage/enable_reqwest_rustls",
   "azure_storage_blobs/enable_reqwest_rustls",
 ]
+google = [
```
Comment (fulmicoton) [inline quickwit/quickwit-storage/Cargo.toml:77]: can we call the feature gfs instead of google?

Comment (Xuanwo) [reply quickwit/quickwit-storage/Cargo.toml:77]: Sure, I will update this place tomorrow. 😋

Comment (fulmicoton) [reply quickwit/quickwit-storage/Cargo.toml:77]: i ll do it actually... i spotted a bunch of other problems.

**Comment by Xuanwo** [comment]
Thanks a lot!

**Comment by fulmicoton** [comment]
Thank you @Xuanwo that's awesome!

**Pull request merged**